srml/session/src/lib.rs

```diff
@@ -115,208 +115,306 @@
+macro_rules! impl_opaque_keys {
+	(
+		pub struct $name:ident ( $( $t:ty ),* $(,)* );
+	) => {
+		impl_opaque_keys! {
+			pub struct $name ( $( $t ,)* );
+			impl OpaqueKeys for _ {}
+		}
+	};
+	(
+		pub struct $name:ident ( $( $t:ty ),* $(,)* );
+		impl OpaqueKeys for _ {
+			$($rest:tt)*
+		}
+	) => {
+		#[derive(Default, Clone, PartialEq, Eq, Encode, Decode)]
+		#[cfg_attr(feature = "std", derive(Debug, Serialize, Deserialize))]
+		pub struct $name($( pub $t ,)*);
+		impl $crate::OpaqueKeys for $name {
+			fn count() -> usize {
+				let mut c = 0;
+				$( let _: $t; c += 1; )*
+				c
+			}
+			fn get_raw(&self, i: usize) -> &[u8] {
+				$crate::count!(impl_opaque_keys (!! self i) $($t),*);
+				&[]
 			}
+			$($rest)*
 		}
+	};
+	( !! $self:ident $param_i:ident $i:tt) => {
+		if $param_i == $i { return $self.$i.as_ref() }
 	}
 }
 
-for_each_tuple!(impl_session_change);
-
-pub trait Trait: timestamp::Trait + consensus::Trait {
-	/// Create a session key from an account key.
-	type ConvertAccountIdToSessionKey: Convert<Self::AccountId, Option<Self::SessionKey>>;
+pub trait ShouldEndSession<BlockNumber> {
+	fn should_end_session(now: BlockNumber) -> bool;
+}
 
-	/// Handler when a session changes.
-	type OnSessionChange: OnSessionChange<Self::Moment>;
+pub struct PeriodicSessions<
+	Period,
+	Offset,
+>(PhantomData<(Period, Offset)>);
+
+impl<
+	BlockNumber: Rem<Output=BlockNumber> + Saturating + Zero,
+	Period: Get<BlockNumber>,
+	Offset: Get<BlockNumber>,
+> ShouldEndSession<BlockNumber> for PeriodicSessions<Period, Offset> {
+	fn should_end_session(now: BlockNumber) -> bool {
+		((now.saturating_sub(Offset::get())) % Period::get()).is_zero()
+	}
+}
 
-	/// The overarching event type.
-	type Event: From<Event<Self>> + Into<<Self as system::Trait>::Event>;
+pub trait OnSessionEnding<AccountId> {
+	/// Handle the fact that the session is ending, and optionally provide the new validator set.
+	fn on_session_ending(i: SessionIndex) -> Option<Vec<AccountId>>;
 }
 
-decl_module! {
-	pub struct Module<T: Trait> for enum Call where origin: T::Origin {
-		fn deposit_event<T>() = default;
+/// Handler for when a session keys set changes.
+pub trait SessionHandler<AccountId> {
+	/// Session set has changed; act appropriately.
+	fn on_new_session<Ks: OpaqueKeys>(changed: bool, validators: &[(AccountId, Ks)]);
 
-		/// Sets the session key of a validator (function caller) to `key`.
-		/// This doesn't take effect until the next session.
-		fn set_key(origin, key: T::SessionKey) {
-			let who = ensure_signed(origin)?;
-			// set new value for next session
-			<NextKeyFor<T>>::insert(who, key);
-		}
+	/// A validator got disabled. Act accordingly until a new session begins.
+	fn on_disabled(validator_index: usize);
+}
 
-		/// Set a new session length. Won't kick in until the next session change (at current length).
-		fn set_length(#[compact] new: T::BlockNumber) {
-			<NextSessionLength<T>>::put(new);
-		}
+pub trait OneSessionHandler<AccountId> {
```

Comment:
```suggestion
pub trait OnSessionHandler<AccountId> {
```

Reply:
It's actually meant to be `OneSessionHandler` (as in "one session handler of many")

Reply:
That's really confusing, since _every other event handler in substrate_ is called `OnX`.

Reply:
It's not a **single** event handler. its counterpart, `SessionHandler` also doesn't handle a **single** event, and therefore the naming needs to be a little more creative.